### PR TITLE
String Slice and String Length

### PR DESCRIPTION
**Pull request by danking**
Caitlin uses these three in a pipeline she's trying to run, so this should get her a little closer to being 100% IR.

I added three new IR pieces:
 - `str` on ints and floats (implemented using Scala string method)
 - `StringSlice` IR node corresponding to java.lang.String.substring (the python-like negative indexing is implemented in the function registry)
 - `StringLength` IR node to calculate number of bytes present in the String (in particular, this incorrectly reports length 4 for the poop emoji)

I added two suites to test these new IR nodes.

### CODE REVIEW

File: src/main/scala/is/hail/expr/ir/Interpret.scala [hunk 1]
```diff
@@ -408,6 +408,21 @@ object Interpret {
...
+        else {
+          val vs = maybeString.asInstanceOf[String]
+          if (vstart < 0 || vstart == vs.length || vstart > vend || vend > vs.length)
+            fatal(s"""string slice out of bounds or invalid: "$vs"[$vstart:$vend]""")
+          else
+            vs.substring(vstart, vend)
+        }
+      case StringLength(s) =>
+        val vs = interpret(s).asInstanceOf[String]
+        if (vs == null) null else vs.getBytes().length
```
Comment (jigold) [inline src/main/scala/is/hail/expr/ir/Interpret.scala]: Shouldn't this just be `vs.length()`?

File: src/test/scala/is/hail/TestUtils.scala [hunk 2]
```diff
@@ -23,11 +23,12 @@ object TestUtils {
...
+    val msg =
+      s"""expected fatal exception with pattern `$regex'
+         |  Found: ${ thrown.getMessage } """
     if (!p)
-      println(
-        s"""expected fatal exception with pattern `$regex'
-           |  Found: ${ thrown.getMessage } """.stripMargin)
-    assert(p)
+      println(msg)
+    assert(p, msg)
```
Comment (jigold) [inline src/test/scala/is/hail/TestUtils.scala]: Doesn't assert print the message if `p` is false?

File: src/test/scala/is/hail/expr/ir/StringSliceSuite.scala [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import org.scalatest.testng.TestNGSuite
+import org.testng.annotations.Test
+
+class StringSliceSuite extends TestNGSuite {
+  @Test def zeroToLengthIsIdentity() {
+    assertEvalsTo(StringSlice(Str("abc"), I32(0), I32(3)), "abc")
+  }
+
+  @Test def outOfBoundsFatals() {
+    assertFatal(StringSlice(Str("abc"), I32(4), I32(4)),
```
Comment (jigold) [inline src/test/scala/is/hail/expr/ir/StringSliceSuite.scala]: Might be worth seeing what Patrick did with the ArraySuite and SetSuites in #3554. He added some nice features.

File: src/test/scala/is/hail/expr/ir/StringSliceSuite.scala [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  @Test def bothSideSliceMatchesIntuition() {
+    assertEvalsTo(invoke("[*:*]", Str("abc"), I32(0), I32(2)), "ab")
+    assertEvalsTo(invoke("[*:*]", Str("abc"), I32(1), I32(3)), "bc")
+  }
+
+  @Test def leftSliceIsPythony() {
+    assertEvalsTo(invoke("[*:]", Str("abc"), I32(-1)), "c")
+    assertEvalsTo(invoke("[*:]", Str("abc"), I32(-2)), "bc")
```
Comment (jigold) [inline src/test/scala/is/hail/expr/ir/StringSliceSuite.scala]: Add tests for -6 etc. large out of bounds indices.

File: src/main/scala/is/hail/expr/ir/Emit.scala [hunk 5]
```diff
@@ -717,6 +717,42 @@ private class Emit(
...
+            Code._empty)
+          )
+        val sliced = region.appendStringSlice(
+          region, vs, vstart, vend - vstart)
+        strict(Code(checks, sliced), cs, cstart, cend)
+
+      case StringLength(s) =>
+        val t = coerce[TString](s.typ)
+        val cs = emit(s)
+        strict(TString.loadLength(region, coerce[Long](cs.v)), cs)
```
[the remark this answers is not in the document]
Comment (danking) [reply src/main/scala/is/hail/expr/ir/Emit.scala]: `strict` will say that this code is missing if `cs` is missing.

File: src/main/scala/is/hail/expr/ir/Interpret.scala (hunk 1 above)
Comment (danking) [reply src/main/scala/is/hail/expr/ir/Interpret.scala]: I don't know what to do exactly. In the region value we represent strings as arrays of bytes, so, for example, the poop emoji is length 4 because it needs two UTF-16 code points (each 16 bits). `String.length` is a very smart function that actually knows how unicode works.

File: src/test/scala/is/hail/TestUtils.scala (hunk 2 above)
Comment (danking) [reply src/test/scala/is/hail/TestUtils.scala]: what was happening was that the message got printed to the stdout, but I'm using SBT to run tests (it re-compiles the relevant files and runs the test suites for this PR in 6 seconds, so it's really nice) and SBT shows me a short summary of what went wrong with only the assert message. So it was really frustrating to see:
```
[info] - bothSidesSliceFunctionOutOfBoundsNotFatal *** FAILED ***
[info]   is.hail.utils.AssertionError: p was false
```
and then I had to scroll up to see the output.

This makes the message something like:

```
[info] - bothSidesSliceFunctionOutOfBoundsNotFatal *** FAILED ***
[info]   t.valuesSimilar(c, expected, t.valuesSimilar$default$3, t.valuesSimilar$default$4) was false , ab (TestUtils.scala:419)
```

which still isn't great but it at least has information about the actual value and the expected value.

Comment (danking) [reply src/test/scala/is/hail/TestUtils.scala]: Ideally the message would be 
```
[info] - bothSidesSliceFunctionOutOfBoundsNotFatal *** FAILED ***
[info]   expected "��" to equal "ab" (TestUtils.scala:419)
```
But I wasn't sure how to make assert not show that useless extra information (I didn't look very hard either though).

File: src/test/scala/is/hail/expr/ir/StringSliceSuite.scala (hunk 4 above)
Comment (danking) [reply src/test/scala/is/hail/expr/ir/StringSliceSuite.scala]: I added the out of bounds tests down below in a separately named function

**Comment by danking** [comment]
ping @jigold would be stellar if you can look at it before end of day today :3

File: src/main/scala/is/hail/asm4s/package.scala [hunk 6]
```diff
@@ -164,7 +164,7 @@ package object asm4s {
     def newArray() = new TypeInsnNode(ANEWARRAY, iname)
   }
 
-  object HailClassLoader extends ClassLoader {
+  object HailClassLoader extends ClassLoader(this.getClass.getClassLoader) {
```
Comment (jigold) [inline src/main/scala/is/hail/asm4s/package.scala]: Why was this needed?

File: src/test/scala/is/hail/expr/ir/StringSliceSuite.scala [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  @Test def isStrict() {
+    assertEvalsTo(StringSlice(NA(TString()), I32(0), I32(2)), null)
+    assertEvalsTo(StringSlice(NA(TString()), I32(-5), I32(-10)), null)
+  }
+
+  @Test def stringIndexMatchesIntuition() {
+    assertFatal(invoke("[]", Str("abc"), I32(-5)),
+      "string slice out of bounds or invalid: \"abc\"\\[-2:-1\\]")
```
Comment (jigold) [inline src/test/scala/is/hail/expr/ir/StringSliceSuite.scala]: I find these error messages confusing since they're translated from what the user entered `[-5]`.

File: src/main/scala/is/hail/asm4s/package.scala (hunk 6 above)
Comment (danking) [reply src/main/scala/is/hail/asm4s/package.scala]: bad commit, removing

File: src/test/scala/is/hail/expr/ir/StringSliceSuite.scala (hunk 7 above)
Comment (danking) [reply src/test/scala/is/hail/expr/ir/StringSliceSuite.scala]: I removed string ref. I'll make a separate PR that handles it with better error messages
